tuf/data/roles.go

```diff
@@ -101,6 +113,94 @@ type Role struct {
 	Email            string   `json:"email,omitempty"`
 }
 
+// BaseRole is an internal representation of a root/targets/snapshot/timestamp role, with its public keys included
+type BaseRole struct {
+	Keys      map[string]PublicKey `json:"keys"`
+	Name      string               `json:"name"`
+	Threshold int                  `json:"threshold"`
+}
+
+// ListKeys retrieves the public keys valid for this role
+func (b BaseRole) ListKeys() KeyList {
+	return listKeys(b.Keys)
+}
+
+// ListKeyIDs retrieves the list of key IDs valid for this role
+func (b BaseRole) ListKeyIDs() []string {
+	return listKeyIDs(b.Keys)
+}
+
+// DelegationRole is an internal representation of a delegation role, with its public keys included
+type DelegationRole struct {
+	BaseRole
+	Paths            []string `json:"paths,omitempty"`
+	PathHashPrefixes []string `json:"path_hash_prefixes,omitempty"`
+}
+
+func listKeys(keyMap map[string]PublicKey) KeyList {
+	keys := KeyList{}
+	for _, key := range keyMap {
+		keys = append(keys, key)
+	}
+	return keys
+}
+
+func listKeyIDs(keyMap map[string]PublicKey) []string {
+	keyIDs := []string{}
+	for id := range keyMap {
+		keyIDs = append(keyIDs, id)
+	}
+	return keyIDs
+}
+
+// RestrictChild restricts the paths and path hash prefixes for the passed in delegation role,
+// returning a copy of the role with validated paths as if it was a direct child
+func (d DelegationRole) RestrictChild(child DelegationRole) (DelegationRole, error) {
```

Comment:
Can probably just be `Restrict`. Makes the function signature (and possibly callers) read better: `parent.Restrict(child)`